---
title: "Illumina Kraken 2 Analysis"
author:
  - name: Simon Grimm
    url: {simongrimm.com}
date: 2022-11-14
description: |
categories:
draft: false
output:
  bookdown::html_document2:
    toc: yes
    toc_float: yes
    toc_collapsed: yes
    df_print: paged
    code_folding: show
    theme: lumen
    warning: false
editor_options:
  markdown:
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  cache = TRUE,
  warning=FALSE,
  autodep = TRUE,
  cache.comments = FALSE,
  engine.opts = list(bash = "-l"),
  dpi = 300
)
```

# Introduction

Here we briefly import, format, and clean kraken2 results of our Illumina data.

# Setup

## R packages

We start by loading all required R packages:

```{r}
# set of data analysis and viz tools
library(tidyverse)

# file system helpers
library(fs)
library(googlesheets4)

# specifying locations within a project
library(here)

# microbiome analysis helpers
library(biomformat)
library(speedyseq)
library(vegan)

# plotting helpers
library(cowplot)
library(patchwork)
library(ggbeeswarm)
```

## Defining directories

```{r}
biom_data_path <- here("data/czid")
metadata_data_path <- here("data/metadata")
kraken_data_path <- here("data/kraken_reports")
output_data_path <- here("output")
```

## Data import

### Metadata
We first read in the metadata for all samples. This table is in a manually curated google sheet, available under this [link](https://docs.google.com/spreadsheets/d/1a7BHgb0_NPgey61IHlsF-RP2_EBhAUyW4e377oHQs1k/edit):

```{r}
#Reading in metadata from this google sheet link:
sample_metadata <- read_sheet(
  'https://docs.google.com/spreadsheets/d/1a7BHgb0_NPgey61IHlsF-RP2_EBhAUyW4e377oHQs1k/edit') %>%
  janitor::clean_names() %>%
  glimpse
```

### Kraken2 reports
The code below reads in each kraken report into an R data table. This code could be made more efficient by using a loop which I will add later.

We also assign column names to the data tables.

 - clade_coverage: Percentage of fragments covered by the clade rooted at this taxon
 - fragments_covered: Number of fragments covered by the clade rooted at this taxon
 - fragments_assigned: Number of fragments assigned directly to this taxon
 - rank_code: A rank code indicating (U)nclassified, (R)oot, (D)omain, (K)ingdom, (P)hylum, (C)lass,
(O)rder, (F)amily, (G)enus, (S)pecies. Taxa that are not any of these 10 ranks have a
rank code that is formed by using the rank code of the closest ancestor rank with a
number indicating the distance from that rank. Ex. Lampyrinae has a rank code of
“F1” because it is a subfamily one step below Lampyridae (the firefly family).
 - ncbi_tax_id: The Taxonomic ID number from NCBI
 - scientific_name: Indented Scientific Name

```{r}
col_names <- c('clade_coverage',
              'fragments_covered',
              'fragments_assigned',
              'rank_code',
              'ncbi_tax_id',
              'scientific_name'
              )

report_10330 <- read.table(
  path(kraken_data_path, "10330_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10170 <- read.table(
  path(kraken_data_path, "10170_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10172 <- read.table(
  path(kraken_data_path, "10172_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10175 <- read.table(
  path(kraken_data_path, "10175_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10176 <- read.table(
  path(kraken_data_path, "10176_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10180 <- read.table(
  path(kraken_data_path, "10180_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10330 <- read.table(
  path(kraken_data_path, "10330_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10331 <- read.table(
  path(kraken_data_path, "10331_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10332 <- read.table(
  path(kraken_data_path, "10332_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10333 <- read.table(
  path(kraken_data_path, "10333_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10334 <- read.table(
  path(kraken_data_path, "10334_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10335 <- read.table(
  path(kraken_data_path, "10335_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
report_10336 <- read.table(
  path(kraken_data_path, "10336_kraken2_report.txt"), sep="\t", col.names = col_names, header=FALSE)
```

One thing I could do: Filter down to species level. Use ncbitaxid as feature identifier. Use second column as count.

pivot wider.

feature id, sample id, abundance.

1. as dataframe
2. assign first column as row names
3. dataframe


Other option: use cbind function.

Another option left_join

Each row is the file name.

Look into metacal function called build_matrix (look at source code).

as(matrix)

Matrix has one limitation, every variable has to have the same type.



In our work here we want to end up with an abundance table. This table has as its first row an OTU (either ncbi_tax_id, or species name, and the individual abundances per sample as the following columns. For reference here is the otu_table of our CZID data:
```{r}
ps <- readRDS(here(output_data_path, "2022_12_08_illumina_phyloseq.Rds"))

otu_czid <- ps %>%
    otu_table()
```

To get where we want we first need to take a look at how our kraken2 dataframes look like.

```{r}
report_10330  %>%
   glimpse()
```

From this table we can see that we have a lot of information that we don't need. We only need the ncbi_tax_id and the fragments_assigned of all species. We can filter down to these columns using the select function and filter down to the species level using the filter function.

**ACTION** The final line of code in the function below should rename the fragments_covered column to the sample name. Using deparse(substitute(.))
```{r}
kraken_reports <- list(report_10170, report_10172, report_10175,
                    report_10176, report_10180, report_10330,
                    report_10331, report_10332, report_10333,
                    report_10334, report_10335, report_10336
                    )

# Creating a function that filters down to species level and selects the ncbi_tax_id and fragments_covered columns
fun_subset_otu_table <- function(x){
  new_column_name <- deparse(substitute(x))
  test <- x %>%
    filter(rank_code == "S") %>%

    select(ncbi_tax_id, fragments_covered) %>%
    # Renaming the fragments_covered column to the sample name
    rename(new_column_name = fragments_covered)
}
```

```{r}
kraken_reports_new_columns <- lapply(kraken_reports, fun_rename_column_to_sample_name)

```



Applying the function to each report in kraken_reports, creating a new list of reports called kraken_reports_subset.
```{r}
kraken_reports_subset <- lapply(kraken_reports, fun_subset_otu_table)
```

Comparing the first report in kraken_reports_subset to the first report in kraken_reports:
```{r}

kraken_reports_subset[1] %>% glimpse()

kraken_reports[1] %>% glimpse()
```

```{r}
otu_kraken <- Reduce(
  function(x, y) full_join(x, y, by = 'ncbi_tax_id', all = TRUE),
  kraken_reports_subset
)

otu_kraken_new_columns <-
  function(x, y) full_join(x, y, by = 'ncbi_tax_id', all = TRUE),
  kraken_reports_new_columns
)
```


```{r}
species_10170 <- report_10170 %>%
  filter(rank_code == "S") %>%
  select(ncbi_tax_id, fragments_covered) # %>%
  # rename the fragments_covered column to the sample name
  col_names(.)
  # set the first column as row names
  # column_to_rownames(., var = "ncbi_tax_id")

species_10172 <- report_10172 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10175 <- report_10175 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10176 <- report_10176 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10180 <- report_10180 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10330 <- report_10330 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10331 <- report_10331 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10332 <- report_10332 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10333 <- report_10333 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10334 <- report_10334 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10335 <- report_10335 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)

species_10336 <- report_10336 %>%
  filter(rank_code == "S")  %>%
  select(ncbi_tax_id, fragments_covered)
```


Now we need to merge all of these dataframes into one. We can do this using the cbind function.

```{r}


```
Now we use the Reduce function to successively apply a function to the elements of a list.
In a more general example this looks the following way:
l_1 = f(v1 ,v2)
l_2 = f(l_1, v_3)

```{r}
# Creating otu_kraken by successively applying the full_join function each ssubs
otu_kraken <- Reduce(
  function(x, y) full_join(x, y, by = 'ncbi_tax_id', all = TRUE),
  species_list)

otu_kraken <- otu_kraken %>%
  column_to_rownames(., var = "ncbi_tax_id")
```